matlab_simulation/05-estimation/EKF_Localization/ekf_localization_v2.m

```diff
@@ -0,0 +1,210 @@
+selectedFeature = zeros(2,1);
+featureFoundFlag=0; %Switches to 1 when feature is found. Used to exit for loop
+
+%% Main loop
+for t=2:length(T)
+    %% Simulation
+    % Select a motion disturbance
+    motionDisturbance = eigenvectorR*sqrt(eigenvalueR)*randn(nStates,1);
+    % Update state with prediction
+    x(:,t) = motion_model(x(:,t-1),u(:,t)) + motionDisturbance;
+    % Take measurement
+    [p,featureInViewFlag] = get2dpointmeasurement(featureMap,x(:,t),RANGE_MAX,THETA_MAX,Q,MEASUREMENT_TYPE);
+    y(1:length(p),t)=p;
+    
+    %% Extended Kalman Filter
+    % Call EKF function
+    % Find first feature in view
+    index = find(featureInViewFlag==1,1);
+    % Determine coordinates of feature used for localization
+    selectedFeature = featureMap(index,:);
+    % Localize with respect to chosen feature
+    [ mu, S, mup, K, I ] = multisensorekf(selectedFeature, MEASUREMENT_TYPE, ...
+        index, mu,u(:,t), S, y(:,t), motion_model, linearized_motion_model, Q, R );
+
+    % Store results
+    mup_S(:,t) = mup;
+    mu_S(:,t) = mu;
+    Inn(:,t) = I;
+    %% Plot results
+    %Plotting initializations
+    figure(1);clf; hold on;
+    axis equal
+    axis([-4 6 -1 7]) % Set to maximum of robot trajectory and map features
+    
+    % Plot states
+    plot(x(1,1:t),x(2,1:t), 'ro--')
+    % Plot map features
+    plot(featureMap(:,1),featureMap(:,2),'go', 'MarkerSize',10,'LineWidth',2);
+    
+    % Plot initial prediction (model propgation)
+    plot(mup_S(1,1:t),mup_S(2,1:t), 'go--')
+    % Plot updated prediction
+    plot(mu_S(1,1:t-1),mu_S(2,1:t-1), 'bx--')
+    
+    %Plot error ellipses
+    mu_pos = [mu(1) mu(2)];
+    S_pos = [S(1,1) S(1,2); S(2,1) S(2,2)];
+    error_ellipse(S_pos,mu_pos,0.95);
+    error_ellipse(S_pos,mu_pos,0.999);
+    
+    % Plot robot measurement to inview features
+    for i = 1:nFeatures
+        if (featureInViewFlag(i))
+            plot(featureMap(i,1),featureMap(i,2),'mx', 'MarkerSize',10,'LineWidth',2)
+            if (MEASUREMENT_TYPE==1) 
+                circle(1,x(1:2,t), y(i,t)); 
+            end
+            if (MEASUREMENT_TYPE==2) 
+                plot([x(1,t) x(1,t)+10*cos(y(i,t)+x(3,t))],...
+                    [ x(2,t) x(2,t)+10*sin(y(i,t)+x(3,t))], 'c');
+            end
+            if (MEASUREMENT_TYPE==3) 
+                plot([x(1,t) x(1,t)+y(2*(i-1)+1,t)*cos(y(2,t)+x(3,t))],...
+                    [ x(2,t) x(2,t)+y(2*(i-1)+1,t)*sin(y(2*i,t)+x(3,t))], 'c');
+            end
+        end
+    end
+    drawnow;
+    
+    if (makemovie) writeVideo(vidObj, getframe(gca)); end
+end
+
+% Add plotting options for figure 1
+title('Range & Bearing Measurements for Localization')
+xlabel('X Postion [Unit]')
+ylabel('Y Position [Unit]')
+legend('Robot State','Map Feature','Initial State Prediction',...
+    'Updated State Prediction')
+
+%Plot measurement vs prediction error time series
+figure(2);clf; hold on;
+plot(T,Inn)
+title('Error Between Measurement and Prediction')
```

Comment:
This error between actual measurement and the expected measurement in EKF is called innovation for some reason (Estimation I pg59)